File: src/opttdev.f
      subroutine opttdev(veg,numplt,numspc,clusid,numcls,
     +                  maxitr,minsiz,sums,numitr,
     +                  relsum,colsum,spcsum,tmpclu)
c
c* passed in/out
c
      integer numplt
      integer numspc
      integer maxitr
      integer numcls
      integer minsiz
      integer numitr
      integer clusid(numplt)
      double precision veg(numplt,numspc)
      double precision sums(0:maxitr)
      double precision relsum(numcls)
      double precision colsum(numcls)
      double precision spcsum(numspc)
c
c* local
c
      integer tmpclu(numplt)
      double precision base
      double precision totdev
      integer clstab(numcls)
      integer moveto
      integer mvefrm
      integer plttrn
      integer flag
c
c********************* one ******************************
c
      call clcdev(veg,numplt,numspc,clusid,numcls,totdev,
     +            relsum,colsum,spcsum)
      base = totdev
      numitr = maxitr
      sums(0) = totdev
      do 18 i=1,numcls
      clstab(i) = 0
   18 continue
      do 17 i=1,numplt
      clstab(clusid(i)) = clstab(clusid(i)) + 1
   17 continue
c
      do 19 k=1,maxitr
      flag = 0
        do 10 i=1,numplt
          do 11 j=1,numplt
          tmpclu(j) = clusid(j)
  11      continue
c
          do 12 j=1,numcls
          if (j .eq. clusid(i)) goto 12
          if (clstab(clusid(i)) .le. minsiz) goto 12
          tmpclu(i) = j
          call clcdev(veg,numplt,numspc,tmpclu,numcls,totdev)
          if (totdev .lt. base) then
            base = totdev
            moveto = tmpclu(i)
            mvefrm = clusid(i)
            plttrn = i
            flag = 1
          endif
   12     continue
   10   continue
      if (flag .eq. 1) then
        clusid(plttrn) = moveto
        clstab(moveto) = clstab(moveto) + 1
        clstab(mvefrm) = clstab(mvefrm) - 1
        sums(k) = base
      else
        numitr = k 
        return
      endif
   19 continue
c
      return
c
      end

File: R/opttdev.R
opttdev <- function (veg,clustering,maxitr=100,minsiz=5) 
{
    if (inherits(clustering, c("partana", "clustering", "partition"))) 
        clustering <-  clustering$clustering
    if (is.numeric(clustering)) {
        if (min(clustering)< 0 || (length(table(clustering)) != max(clustering))) {
            cat('WARNING: renumbering clusters to consecutive integers\n')
            clustering <- match(clustering,sort(unique(clustering)))
        }
    }

    numplt <- nrow(veg)
    numspc <- ncol(veg)
    numcls <- length(table(clustering))
    sums <- rep(0,maxitr+1)
    numitr <- 0
    relsum <- rep(0,numcls)
    colsum <- rep(0,numcls)
    spcsum <- rep(0,numspc)
    tmpclu <- rep(0,numplt)
    tmp <- .Fortran('opttdev',
                    as.double(as.matrix(veg)),
                    as.integer(numplt),
                    as.integer(numspc),
                    clusid = as.integer(clustering),
                    as.integer(numcls),
                    as.integer(maxitr),
                    as.integer(minsiz),
                    sums = as.double(sums),
                    numitr = as.integer(numitr),
                    as.double(relsum),
                    as.double(colsum),
                    as.double(spcsum),
                    as.integer(tmpclu),
                    PACKAGE='optpart')
    out <- list()
    out$dev<- tmp$sums[1:tmp$numitr]
    out$numitr = tmp$numitr
    out$clustering <- tmp$clusid
    class(out) <- c('opttdev','clustering')
    return(out)
}
